### ISSUE TITLE
audio function does not work

### ISSUE THREAD

**Issue by pangguoming** [youtube/yt-watchme#23]
hey @vshcryabets, did you find that audio function does not work? did you find a way to fix it? if you did it please let me know. thanks.,

**Comment by danman81** [comment]
Have you gotten this fixed, @pangguoming?  If you take a look near the bottom of this thread:

https://github.com/youtube/yt-watchme/issues/16

You can see what is might be wrong with the audio is a depricated call to avcodec_encode_audio in avcodec.h, which should be calling avcodec_encode_audio2 instead.  I've tried compiling ffmpeg on my own, as well as using precompiled libraries provided by @garyjamessilva at the above thread.  Hope this helps u!

**Comment by Lightartist** [comment]
Hey everybody, is someone able to solve the audio problem? If it is complicated or takes too much time, we will pay for your service! Thanks in advance for your message!

**Comment by pangguoming** [comment]
@danman81 thanks a lot for your suggestion, i will try to fix it now.

if i have any progress , i will tell you @Lightartist

**Comment by BenGroot** [comment]
@pangguoming thanks! I hope you'll be able to fix it!

**Comment by danman81** [comment]
thanks, @pangguoming! let us know how it goes!

**Comment by danman81** [comment]
hey, @pangguoming have you made any progress?  let me know if you run into errors and so i can help you debug!

**Comment by pangguoming** [comment]
@danman81  sorry, i am occupied in other important work, i did not fix the audio function yet. i will continue to try to fix it in my free time

**Comment by danman81** [comment]
Thank you for the update, @pangguoming !  please let us know if you make more progress!  and feel free to post any errors you run into!

**Comment by GavinCaptain** [comment]
@pangguoming , @danman81 :
I got the audio part worked!
The points are as follows:

1. in the file "ffmpeg-jni.c", in the function 'Java_com_acer_camera_pip_Ffmpeg_init'.
Check this code:
    audio_codec_ctx->channels = 2; 

If your device only support mono, you need to set this to 1.
Also, in the file "AudioFrameGrabber.java", in the method "recordThread()", 
Check this code:
    int channelConfiguration = AudioFormat.CHANNEL_CONFIGURATION_STEREO;
you need to set it to AudioFormat.CHANNEL_CONFIGURATION_MONO.

**Comment by danman81** [comment]
@gavincaptain I love you from the bottom of my heart!  I will try this tonite and let u know how it goes!

**Comment by pangguoming** [comment]
@gavincaptain i love you deeply, did you get married?

**Comment by pangguoming** [comment]
@gavincaptain  chould you please share your whole project ?

**Comment by danman81** [comment]
@GavinCaptain just wanted to update everyone on my progress.  when I try your optional suggestion #4, I get the following error:

The encoder 'aac' is experimental but experimental codecs are not enabled, add '-strict -2' if you want to use it.

But this was resolved when I added "av_dict_set(&opts, "strict", "experimental", 0);" back in along with "av_dict_set(&opts, "-bsf:a", "aac_adtstoasc", 0);" instead of replacing.

Also, when I try suggestion #2, with or without #4, I get the following error during runtime:

Specified sample format s16 is invalid or not supported

i'm running this on windows, so I did not add "--enable-libfdk_aac" to build.sh. but I can recompile libffmpeg.so without problems.  i also have libfdk-aac.a built.  anyways, when I google the error, looks like someone was able to resolve it by downgrading ffmpeg:

https://bbs.archlinux.org/viewtopic.php?pid=1245854#p1245854

I think i'm using ffmpeg-2.5.3, so maybe that is the problem?  maybe i'll try downgrading to ffmpeg2.3.6 to see if that helps.   

Either way, this is much farther than I've gotten, so thanks so much again!!

**Comment by GavinCaptain** [comment]
@danman81 : 
## 1.
Yes, the level of ffmpeg DOES matter. I also failed when I used higher level ffmpeg. (but I forgot the error message...^^U)

## 2.
"Specified sample format s16 is invalid or not supported" -> Yes, if you don't add '--enable-libfdk_aac', you will encounter this problem.

I think the ffmpeg module is a "big machine", and all the arguments is to tell the machine which 'component' should be used in each part during the compiling process.
Each component may support different audio format.

So, it's ok for compliing with no error, if you do not use ''--enable-libfdk_aac'.
Because you will get the default ffmpeg's audio format support with that "*.a"
So, if you add '--enable-libfdk_aac' and want to set:
__audio_codec_ctx->sample_fmt = AV_SAMPLE_FMT_FLT;__
You will get some error message like "not supported..", too.

## 3.
Don't know if this helps, my whole process is as follows:
1. compile x264, aac module to get .a in Ubuntu
2. compile lame module to get .a in Ubuntu (but I don't think this will be used, because I did not add '--enable-libmp3lame' argument in step 3)
3. compile ffmpeg module to get .a in Ubuntu

4. copy all the .a to the "watchme" project folder
5. Use Eclipse to compile whole project(including the libffmpeg.so) on Window7
[Ps.] I use ndk-r10d both on Windows and Ubuntu

Hope you could make audio work!

**Comment by GavinCaptain** [comment]
@pangguoming : 
Sorry, I cannot share the whole project. Because this is my company's project, not mine. I am afraid of the copyright issue if I do so.
But I will keep answering the problem as possible as I could.

**Comment by pangguoming** [comment]
@GavinCaptain  it does not matter. i will try to do it as you descripted thanks a lot

**Comment by sybercoda** [comment]
I'm having issues getting this app to work as well. I just don't understand how so many changes are required to get it to work when it was just uploaded to github 5 months ago.

**Comment by garyjamessilva** [comment]
Well you will understand what is required to get this working if you start
with only what Google posted to Github originally, they didn't post any of
the ffmpeg & associated libraries (or various other required parts), just a
note saying -

'Include cross-platform compiled streaming libraries. Either libffmpeg.so
under src/main/jniLibs/armeabi, or another streaming library with modifying
VideoStreamingInterface'

Originally Google did not include the JNI code either, but did subsequently
after it was asked for.

In one comment someone said this was like being given a car with no engine,
I thought that was about right as you had to find, assemble (& fix) all
those parts with only few mostly out of date instructions.

**Comment by sybercoda** [comment]
@garyjamessilva I compiled ffmpeg for android, rmtpdump, openssl after jumping through hoops. Im now trying to compile the app with the libraries via Android Studio but I'm getting errors so I think I'm going to try eclipse instead. I hope things are smoother porting this to iOS but that is probably a whole new can of worms. FYI, look Into open source streaming solutions for drones as a viable substitute for this project. Thanks again for your reply.

**Comment by pangguoming** [comment]
i am doing it with windows7 , there is a lot of questions. i had better to do it with ubuntu

**Comment by sybercoda** [comment]
@pangguoming I tried it with windows/android studio at first with no luck as well. Then tried with Ubuntu and was able to compile, much easier! Now I'm going to use eclipse on ubuntu because Android Studio is becoming troublesome. I think Eclipse on Ubuntu is your best bet. Command line compile as much as possible. I still may have to go back and recompile again if audio isn't working, per the above procedure.

**Comment by pangguoming** [comment]
@sybercoda  thanks a lot for your suggestion

**Comment by pangguoming** [comment]
@GavinCaptain could you please share your libffmpeg.so file ? so that i don't have to compile it with windows .

**Comment by GavinCaptain** [comment]
@pangguoming  : 
Here you are. Try it~
https://onedrive.live.com/?cid=eaf2ffc47d5c37b0&id=EAF2FFC47D5C37B0%21720&ithint=file,so&authkey=!AEfiV8h2FkG92Xs

**Comment by GavinCaptain** [comment]
By the way, I found that compiling ffmpeg to .a on Windows is not so easy. 
Lots of error happened..., and I did not succeed.
However, compiling ffmpeg to .a on Ubuntu is much easier.

**Comment by pangguoming** [comment]
@GavinCaptain 
i want to use ubuntu, but my project is created with windows,so.... you konw.

with your libffmpeg.so file , google returned 403 error: stream is inactive. and ffmpeg-ini init function return false.

**Comment by pangguoming** [comment]
@GavinCaptain 
i want to use ubuntu, but my project is created with windows,so.... you konw.

with your libffmpeg.so file , google returned 403 error: stream is inactive. and ffmpeg-ini init function return false.
Could not open file /data/data/com.google.android.apps.watchme/files/outputStreamFile.flv

**Comment by GavinCaptain** [comment]
@pangguoming :
Oh, you need to change the "ffmpeg-jni.c" to meet your own need.
It seems that only share libFFmpeg.so does not work, because it includes all my customized settings.

However, I am not so sure what do you mean "my project is created with windows".
In my case, I import the project 'yt-watchme' with Eclipse on Window7, too.
I only compile ffmpeg on Ubuntu, for getting all *.a which the project 'yt-watchme' needs.
Then I finish all other works with Eclipse on Window7. (Of course, you need to install Cygwin and NDK on Windows first)

**Comment by sybercoda** [comment]
Is there any particular reason or benefit to using ffmpeg for android for live streaming to YouTube vs using the MediaCodec class ? It is to my understanding that MediaCodec offers hardware encoding and is much faster.

**Comment by pangguoming** [comment]
yes , sir, you are right , i mean  i need install JDK, eclipse, ADT softer ware to a ubuntu computer. i am so lazy.

**Comment by pangguoming** [comment]
@GavinCaptain yes , sir, you are right , i mean i need install JDK, eclipse, ADT NDK softer ware etc to an ubuntu computer.  i am so lazy.

**Comment by sybercoda** [comment]
@pangguoming You will also need libssl library on Ubuntu or you will receive an error

**Comment by GavinCaptain** [comment]
@sybercoda :
yes, I am trying to use MediaCodec encoder now, and got stuck....
In short, Android only provides video/audio encoder and muxer for mp4.
However, if we want to live stream to Youtube, we need to produce .flv stream.
So,I'm trying to use ffmpeg for producing .flv with MediaCodec encoder.
But no luck......

**Comment by JavierCabanas** [comment]
@GavinCaptain:
As far as I know, MediaCodec encoder doesn't write the information to make the video readable until the closure of the file. So you won't be able to provide ffmpeg the output data provided by MediaCodec while recording the video

**Comment by pangguoming** [comment]
@GavinCaptain    Thans a lot for your persevere answer.

finally, i have only one question  ,when i compile libffmpeg.so file.
it shows the Error: The encoder 'aac' is experimental but experimental codecs are not enabled, add '-strict -2' if you want to use it.

According to https://trac.ffmpeg.org/wiki/Encode/AAC, i konw  '-strict -2'  is a command parameter
Do you know how to solve that problem ?.

**Comment by pangguoming** [comment]
if i did it , i will share my whole project . to show everybody my gratitude . LOL

**Comment by danman81** [comment]
Thanks again @GavinCaptain ! tonight, I will try to set the --enable-libfdk_aac flag somewhere else in android studio.  it's my ndk-build command that is not picking up that flag.  I see a bunch of other ffmpeg flags in config.h, but i don't think they're being picked up -- i'm sure it's just somewhere else I need to set them.  if that does not work, i will take your suggestion and use linux to compile.  i will also try the .so file you graciously uploaded.

@pangguoming you need to add both lines:  "av_dict_set(&opts, "strict", "experimental", 0);" and  "av_dict_set(&opts, "-bsf:a", "aac_adtstoasc", 0);" to get rid of that error.  hope that helps!

**Comment by pangguoming** [comment]
@danman81  no sir,  i modified the code as you said, but get the error:
I/ffmpeg-jni(7551): Specified sample format s16 is invalid or not supported
I/ffmpeg-jni(7551): audio frame size: 0
I/ffmpeg-jni(7551): Error opening audio codec: -22

**Comment by danman81** [comment]
hey @pangguoming , you need to set the --enable-libfdk_aac flag just like i need to.  i don't think it's a problem with the version, at least not yet...  i probably won't get a chance to look at this today myself, but i will keep u posted!

**Comment by pangguoming** [comment]
@danman81   yes ,sir, i add --enable-libfdk_aac flag to config.h file. but i still get that error :s16 is invalid or not supported

**Comment by danman81** [comment]
very interesting, @pangguoming .. where did you set --enable-libfdk_aac?   it's possible it won't fix the same error for me either then....

**Comment by pangguoming** [comment]
@danman81  first of all , i am not very family with ffmpeg ,
i noticed that a lot of compile flags are in  config.h  file. but i know that file is automatically generated when we compile the whole ffmpeg libiary.  
so i think we need to modify Android.mk

**Comment by pangguoming** [comment]
@danman81 finally, we had better to compile it with ubuntu rather than windows

**Comment by danman81** [comment]
thanks, @pangguoming !  i will try your suggestions tomorrow!

**Comment by GavinCaptain** [comment]
@pangguoming , @danman81 :
I have never compile the *.a on Windows, so the following may only works on Ubuntu.

1. The config.h is auto-generated. 
It's generated by the command __"./configure"__ with many arguments, like "--enable-libfdk_aac"...etc.

2. After doing step 1, the needed config.h is generated. Then we can use the command 'make' to compile
Theoriertically, modify config.h and then do 'make' should also works, too.

3. For the convenience, it's usually merge step 1 and 2 to a script file. In my case, I made a file "build.sh".
The usage is './build.sh'.
Btw, when you compile on Ubuntu, maybe you will got some errors like 'cannot execute'.
You could do ''chmod *.sh 777" or "chmod configure 777" to get the correct right to execute...

**Comment by pangguoming** [comment]
@GavinCaptain   i got it ,thanks

**Comment by sybercoda** [comment]
Anyone know why I'm receiving the errors below when I try to run the app? Note that line 153 of the 012v.c file is as follows "  .long_name      = NULL_IF_CONFIG_SMALL("Uncompressed 4:2:2 10-bit"), " and I'm using  -std=c99 mode in the Android.mk. Also, I'm using the MainActivity build supplied by @garyjamessilva. 

C:\WatchMe\MainActivity\mainActivity\src\main\jni\libavcodec\012v.c
Error:(153, 5) error: initializer element is not constant
Error:(153, 5) error: (near initialization for 'ff_zero12v_decoder.long_name')

**Comment by danman81** [comment]
okay, so i did some reading and found that this ./configure stuff might not be supported in android.mk or something along those lines so i'm building with linux as suggested..  ran into "ERROR: libfdk_aac" not found when i run ./configure, so i did some more reading and found that the libfdk_aac license is not compatible with the gpl, which is why it's not included -- do i need to worry about licensing issues?  here is some more info:

https://trac.ffmpeg.org/wiki/Encode/AAC

**Comment by sybercoda** [comment]
@danman81 Good catch! I also found the discussion below. Sounds like you're prohibited from compiling it with ffmpeg but would instead have to reference it as externally shared library. https://ffmpeg.zeranoe.com/forum/viewtopic.php?t=877

**Comment by GavinCaptain** [comment]
@danman81 : 
Have you tried "--enable-nonfree" argument?

**Comment by danman81** [comment]
@GavinCaptain yes, i have tried the --enable-nonfree argument -- the problem, however, was that I did not install libfdk_aac...  however, the real problem is that i cannot use fdk_aac because it's not free, which would make my code non-distributable, which really defeats the purpose of using ffmpeg in the first place.   i think my solution is going to be to not use fdk_aac, or in other words, not use the AV_SAMPLE_FMT_S16 audio codec, so i will modify AudioFrameGrabber.java instead to not output AudioFormat.ENCODING_PCM_16BIT.  

that being said, @iulukaya, i'm going to go out on a limb and say that i really think Google dropped the ball here -- they should really be helping us to debug their own code... please!

**Comment by garyjamessilva** [comment]
If Google was seriously interested in us using this live streaming facility
properly it would be a one line call from within the Android SDK.

**Comment by danman81** [comment]
hey friends, so it looks like android.media.AudioRecord from AudioFrameGrabber.java only supports ENCODING_PCM_16BIT, as talked about here:

http://stackoverflow.com/questions/4781781/why-does-audiorecord-getminbuffersize-return-error-bad-value-2

so, it looks like audio_codec_ctx->sample_fmt must be set to a 16bit format in ffmpeg-jni.c as @GavinCaptain says -- this is a limitation of android.media.AudioRecord.  there has to be a way to hack this together without using non-free components of ffmpeg!  fix this google!!!

**Comment by danman81** [comment]
hey guys, so i had some thoughts about this while i was sleeping and i realized it may be possible to convert the 16BIT PCM audio stream to 8BIT, programatically.  does anyone else agree with this approach?

**Comment by GavinCaptain** [comment]
@danman81  :
I'm not familiar with audio things. But It sounds reasonable to me.
However, the transform might be another tough work..... Hope you can work it out!

**Comment by paulvaskar** [comment]
Hello ,

I am getting the following error while trying to start live streaming using YouTube live streaming API form my android application. Please help me to resolve the issue.

Cannot open connection tcp://a.rtmp.youtube.com:1935

06-15 20:11:54.863 2109-2109/com.happytv.mobapp I/ffmpeg-jni﹕ ERROR: Could not open file rtmp://a.rtmp.youtube.com/live2/............

Ffmpeg.init() returned false

I am using the precompiled ffmpeg library provided with the watchme application. If there is any issue with the ffmpeg library please provide me a working version of the same which I can use in my application.

Thanks in advance
Vaskar

**Comment by arthurpalves** [comment]
Hi guys, I've been testing this project recently and just now I paid attention to the audio to see it isn't working. @GavinCaptain, I've tried to recompile with `--enable-libfdk_aac` following your suggestions, but that didn't work very well. Could you share your `libffmpeg.so`? Perhaps it would work for me. (The link you shared above doesn't work anymore)

**Comment by achinverma** [comment]
Hello ,

I am getting the following error while trying to start live streaming using YouTube live streaming API form my android application. Please help me to resolve the issue.

Cannot open connection tcp://a.rtmp.youtube.com:1935

06-15 20:11:54.863 2109-2109/com.happytv.mobapp I/ffmpeg-jni﹕ ERROR: Could not open file rtmp://a.rtmp.youtube.com/live2/............

Ffmpeg.init() returned false

**Comment by jerrybi** [comment]
@GavinCaptain hello,GavinCaptain,i have finally compiled the libffmpeg.so according to your suggestions1.enable fdk-aac and x264 and build all the .a in ubuntu 2.generate libffmpeg.so in androidstudio ndk 3.modify ffmpeg-jni.c
now my problem is i can hear some voice on youtube but can not hear clear,seems that there are a lot of datas missing when uploading.i don't know where may be the problem.do you have this problem?

**Comment by pangguoming** [comment]
@jerrybi if you did it finally , please share your code. Thanks a lot

**Comment by SylvainGirod** [comment]
Hi everybody,
I'm having issues with the audio codec too.
As @pangguoming, I get the -22 error message. The problem is that I can't configure and recompile ffmpeg outside my project, because of companies restriction.
As libfdk_aac seems to be compiled with libfdk_aac, but not activated, could someone, please, provide a config.h file generated with the `--enable-libfdk_aac` parameter please?

**Comment by menna-ali** [comment]
@GavinCaptain 

I am trying to follow your solution to compile ffmpeg-2.3.6 for android using the build script you posted. However, I keep getting the error: libfdk_aac not found. Any idea why this might be happening?

**Comment by vranacse** [comment]
Any one got the right solution of the problem ?
@all

**Comment by mustafalml** [comment]
@vranacse I'd recommend using JavaCV to avoid all the ffmpeg hassle. This is how I eventually got the app to work properly. JavaCV is a wrapper library that provides a nice interface so that you don't have to write native code directly or deal with it's issues.

**Comment by menna-ali** [comment]
@vranacse I'd recommend using JavaCV to avoid all the ffmpeg hassle. This is how I eventually got the app to work properly. JavaCV is a wrapper library that provides a nice interface so that you don't have to write native code directly or deal with it's issues.

**Comment by vranacse** [comment]
@menna-ali  would have some sample or something so that , i can work

**Comment by vranacse** [comment]
@jerrybi  can you help

**Comment by menna-ali** [comment]
@vranacse you can download JavaCV from the repo here: https://github.com/bytedeco/javacv and follow the readme file to add the necessary files to your project. This repo also has some samples, look for the RecordActivity.java. This is a good starting point. However, you would need to make a couple of changes in order to make it stream to youtube:
1. change ffmpeg_link value to rtmp link from youtube https://github.com/bytedeco/javacv/blob/master/samples/RecordActivity.java#L104

2. Replace record length with zero so that you have an active stream, else you will get a 403 stream is inactive exception
https://github.com/bytedeco/javacv/blob/master/samples/RecordActivity.java#L148

**Comment by vranacse** [comment]
Yup let me work do this . Thank you

**Comment by vranacse** [comment]
solved it by some other way, need working demo app with both video and audio working ! contact : vranacse@gmail.com

**Comment by MuthuEphron** [comment]
@GavinCaptain  Please share your ffmpeg  jni code or is there any option work  audio in windows

**Comment by MuthuEphron** [comment]
@pangguoming   I am very struggling in audio function please help me in windows in android studio.

**Comment by MuthuEphron** [comment]
@GavinCaptain   Please share your jni code i didn't a hear the audio while live streaming.

**Comment by MuthuEphron** [comment]
@garyjamessilva  I succesfully live streaming but i didn;t  hear the audio. Please let me know you fixed the issues.

**Comment by MuthuEphron** [comment]
@ulukaya  Sir  changed the JNI Code as per  requirements not working .

**Comment by Muthukumar39** [comment]
@GavinCaptain Please share your ffmpeg jni code or is there any option work audio in windows

**Comment by Muthukumar39** [comment]
@pangguoming I am very struggling in audio function please help me in windows in android studio.

**Comment by Muthukumar39** [comment]
@GavinCaptain Please share your jni code i didn't a hear the audio while live streaming.

**Comment by Muthukumar39** [comment]
@garyjamessilva I succesfully live streaming but i didn;t hear the audio. Please let me know you fixed the issues.

**Comment by Muthukumar39** [comment]
@ulukaya Sir changed the JNI Code as per requirements not working .

**Comment by sourav-dev** [comment]
@GavinCaptain I apply your solution and its error free but still audio is not working while broadcasting a live video.
If Any one got the right solution of the problem? please share the solution.

**Comment by ramjanbashashaik** [comment]
@**pangguoming**   Where do I find build.sh file in my windows?

**Comment by sabodh** [comment]
@vranacse Hi, Could you please share the code sample that you have , I am not able to solve the audio problem yet . Please help
